control-plane/subcommand/sync-catalog/command.go

```diff
@@ -393,6 +415,85 @@ func (c *Command) Run(args []string) int {
 	}
 }
 
+// remove all k8s services from Consul.
+func (c *Command) removeAllK8SServicesFromConsulNode(consulClient *api.Client, nodeName string) error {
+	node, _, err := consulClient.Catalog().NodeServiceList(nodeName, &api.QueryOptions{Filter: c.flagFilter})
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	services := node.Services
+	errChan := make(chan error, 1)
+	batchSize := 300
+	maxRetries := 2
+	retryDelay := 200 * time.Millisecond
+
+	// Ask for user confirmation
+	reader := bufio.NewReader(os.Stdin)
+	for {
+		c.UI.Info(fmt.Sprintf("Are you sure you want to delete %v K8S services from %v? (y/n): ", len(services), nodeName))
+		input, _ := reader.ReadString('\n')
+		input = strings.TrimSpace(input)
+		if input == "y" || input == "Y" {
+			break
+		} else if input == "n" || input == "N" {
+			return nil
+		} else {
+			c.UI.Info("Invalid input. Please enter 'y' or 'n'.")
+		}
+	}
+
+	for i := 0; i < len(services); i += batchSize {
+		end := i + batchSize
+		if end > len(services) {
+			end = len(services)
+		}
+
+		wg.Add(1)
+		go func(batch []*api.AgentService) {
+			defer wg.Done()
+
+			for _, service := range batch {
+				err := retryOps(func() error {
+					_, err := consulClient.Catalog().Deregister(&api.CatalogDeregistration{
+						Node:      nodeName,
+						ServiceID: service.ID,
+					}, nil)
+					return err
+				}, maxRetries, retryDelay, c.logger)
+				if err != nil {
+					if len(errChan) == 0 {
+						errChan <- err
+					}
+				}
+			}
+			c.UI.Info(fmt.Sprintf("Processed %v K8S services from %v", len(batch), nodeName))
+		}(services[i:end])
+		wg.Wait()
+	}
+
+	close(errChan)
+	if err = <-errChan; err != nil {
+		return err
+	}
```

Comment:
I find the `errgroup` package ([docs](https://pkg.go.dev/golang.org/x/sync/errgroup#Group)) super helpful for this kind of pattern. It's basically a `WaitGroup` plus error handling

```go
var eg errgroup.Group
for ... {
	eg.Go(func() error {
		err := doSomething()
		return err
	})
}

// eg.Wait() will wait for all of the goroutines above to complete
// or for one of them to return an error, whichever happens first
if err := eg.Wait(); err != nil {
	return err
}
```

Reply:
Thanks for the reference and the attached example. There is a little nuance here.

- The tasks are divided into batches and executed batch by batch in sequence.
- Since the tasks executed here are not critical, in each batch, they are allowed to fail and the `errChan` only collects the first failure.
- By the end of the whole tasks, it just highlights to the operator that not all of services were successfully de-registered and here is one example failure error.

This means, even though we introduced the errgroup, we still need to keep the `errChan` to collect the failure(The reason I didn't collect all failure is that the number of services is potentially a large number) and we won't gain many benefits for doing this.

Reply:
The `<Group>.Wait()` function does what you're describing for the channel I believe -- it returns the first non-nil error of any function that was run in the group ([docs](https://pkg.go.dev/golang.org/x/sync/errgroup#Group.Wait))

Reply:
Yeah right, the tricky part here is that, <Group>.Go() execute the batches in parallel, but here we added work group to force the batches to be executed in sequence.

Reply:
Updated the code to execute the batch in sequence but execute the tasks inside batch concurrently